Rs/intro (#121)
neuralmagic/docs: src/content/user-guide/deploying-deepsparse/aws-lambda.mdx

---
title: "AWS Lambda"
metaTitle: "Using DeepSparse on AWS Lambda"
metaDescription: "Deploy DeepSparse in a Serverless framework with AWS Lambda"
index: 3000
---

# Deploying with DeepSparse on AWS Lambda

AWS Lambda is an event-driven, serverless computing infrastructure for deploying applications at minimal cost. Since
DeepSparse runs on commodity CPUs, you can deploy DeepSparse on Lambda!

The DeepSparse GitHub repo contains a [guided example](https://github.com/neuralmagic/deepsparse/tree/main/examples/aws-lambda) 
for deploying a DeepSparse Pipeline on AWS Lambda for the sentiment analysis task.

The scope of this application encompasses:
1. The construction of a local Docker image.
2. The creation of an ECR repo in AWS.
3. Pushing the local image to ECR.
4. The creation of the appropriate IAM permissions for handling Lambda.

> 5. The creation of a Lambda function alongside an API Gateway in a Cloudformation stack. 

Reviewer: ```suggestion
5. The creation of a Lambda function alongside an API Gateway in a CloudFormation stack. 
```

## Requirements
The following credentials, tools, and libraries are also required:
* The [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/getting-started-install.html) version 2.X that is [configured](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html). Double check if the `region` that is configured in your AWS CLI matches the region passed in the SparseLambda class found in the `endpoint.py` file. Currently, the default region being used is `us-east-1`.
* The AWS Serverless Application Model [(AWS SAM)](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/what-is-sam.html), an open-source CLI framework used for building serverless applications on AWS.
* [Docker and the `docker` cli](https://docs.docker.com/get-docker/).
* The `boto3` python AWS SDK: `pip install boto3`.

## Quick Start

```bash 
git clone https://github.com/neuralmagic/deepsparse.git
cd deepsparse/examples/aws-lambda
pip install -r requirements.txt
```
## Model Configuration

To use a different sparse model please edit the model zoo stub in the `Dockerfile`. 

> To change pipeline configuration (i.e., change task, engine etc.), edit the pipeline object in the `app.py` file. Both files can be found in the `/lambda-deepsparse/app` directory.

Reviewer: ```suggestion
To change pipeline configuration (e.g., change task, engine), edit the pipeline object in the `app.py` file. Both files can be found in the `/lambda-deepsparse/app` directory.
```

## Create Endpoint

Run the following command to build your Lambda endpoint.

```bash
python endpoint.py create
```
## Call Endpoint

> After the endpoint has been staged (~3 minute), AWS SAM will provide your API Gateway endpoint URL in CLI. You can start making requests by passing this URL into the LambdaClient object. Afterwards, you can run inference by passing in your text input:

Reviewer: ```suggestion
After the endpoint has been staged (~3 minutes), AWS SAM will provide your API Gateway endpoint URL in CLI. You can start making requests by passing this URL into the LambdaClient object. Afterwards, you can run inference by passing in your text input:
```
